package/kedro_viz/data_access/repositories/modular_pipelines.py

```diff
@@ -83,163 +128,201 @@ def get_or_create_modular_pipeline(
 
-        is_internal_output = modular_pipeline_id in output_node.modular_pipelines
-        if is_internal_output:
-            self.tree[modular_pipeline_id].internal_outputs.add(output_node.id)
-        else:
-            self.tree[modular_pipeline_id].external_outputs.add(output_node.id)
+    def _add_children(self, modular_pipeline_id: str, kedro_nodes: List[Node]):
+        """
+        Add children to a modular pipeline. Here we follow the below rules
+        - A kedro node is added as a child to a modular pipeline if
+          it has the namespace of that modular pipeline
+        - Inputs/Outputs of a kedro node are added as children
+          to the modular pipeline if they are not inputs/outputs
+          of that modular pipeline and are not parameter datasets
+        - Any input/output of a modular pipeline is a child
+          of the parent pipeline unless it is an input/output
+          of that parent pipeline
+        - Any input/output of a top-level modular pipeline is added
+          as a child to the root modular pipeline
 
-    def add_child(self, modular_pipeline_id: str, child: ModularPipelineChild):
-        """Add a child to a modular pipeline.
         Args:
-            modular_pipeline_id: ID of the modular pipeline to add the child to.
-            child: The child to add to the modular pipeline.
-        Example:
-            >>> modular_pipelines = ModularPipelinesRepository()
-            >>> modular_pipeline_child = ModularPipelineChild(
-            ...     id="dataset",
-            ...     type=GraphNodeType.DATA
-            ... )
-            >>> modular_pipelines.add_child("data_science", modular_pipeline_child)
-            >>> data_science_pipeline = modular_pipelines.get_or_create_modular_pipeline(
-            ...     "data_science"
-            ... )
-            >>> assert data_science_pipeline.children == {modular_pipeline_child}
+            modular_pipeline_id (str): The ID of the modular pipeline to add children to.
+            kedro_nodes (List[Node]): The kedro nodes (and it's related datasets/parameters)
+                    to add as children.
         """
         modular_pipeline = self.get_or_create_modular_pipeline(modular_pipeline_id)
-        modular_pipeline.children.add(child)
+        modular_pipeline_inputs_outputs = set(modular_pipeline.inputs) | set(
+            modular_pipeline.outputs
+        )
 
-    def extract_from_node(self, node: GraphNode) -> Optional[str]:
-        """Extract the namespace from a graph node and add it as a modular pipeline node
-        to the modular pipeline repository.
+        for kedro_node in kedro_nodes:
+            # add kedro node as a child to the modular pipeline
+            if kedro_node.namespace == modular_pipeline_id:
+                kedro_node_id = _hash(str(kedro_node))
+                modular_pipeline.children.add(
+                    ModularPipelineChild(id=kedro_node_id, type=GraphNodeType.TASK)
+                )

```

Comment:
Instead of nesting this with 2 extra levels, you could do something like:

```python
children = [n for n in kedro_nodes if n.namespace == modular_pipeline_id]
modular_pipeline.children = [ModularPipelineChild(id=_hash(str(n)), type=GraphNodeType.TASK) for n in children]
```

Much more readable and explainable:
1. We get all the nodes matching exactly the pipeline namespace
2. Add them as children to the pipeline (this one could be `.extend` instead of `=` if there's other places that add to children called before this one)

Two actions, two lines of code...